File: src/main/排序/归并排序/面试题51/ReversePairs.java
package 排序.归并排序.面试题51;

/**
 * @auther: sotowang
 * @date: 2020/4/24 15:16
 */
public class ReversePairs {
    private int count = 0;
    private void merge(int[] nums,int low,int high,int mid){
        int[] tem = new int[high-low+1];
        int left = low;
        int right = mid+1;
        int index = 0;
        while(left<=mid && right<= high){
            if(nums[left]<=nums[right]){
                tem[index++] = nums[left++];
            }else{
                count+=mid-left+1;
                tem[index++] = nums[right++];
            }
        }
        while(left<=mid){
            tem[index++] = nums[left++];
        }
        while(right<=high){
            tem[index++] = nums[right++];
        }
        for(int i=0;i<tem.length;i++){
            nums[low++] = tem[i];
        }
    }
    private void mergeSort(int[] nums,int low,int high){
        if(low>=high){
            return;
        }
        int mid = low+((high-low)>>1);
        mergeSort(nums,low,mid);
        mergeSort(nums,mid+1,high);
        merge(nums,low,high,mid);

    }
    public int reversePairs(int[] nums) {
        if(nums.length == 0){
            return 0;
        }
        mergeSort(nums,0,nums.length-1);
        return count;

    }

    public static void main(String[] args) {
        int[] nums = {7, 5, 6, 4};
        assert new ReversePairs().reversePairs(nums) == 5;
    }
}

File: src/main/排序/归并排序/t493/reversePairs.go
package t493

func reversePairs(nums []int) int {
	if len(nums) < 2 {
		return 0
	}
	count := 0
	mergeSort(nums, 0, len(nums)-1, &count)
	return count
}
func mergeSort(nums []int, start, end int, count *int) {
	if start == end {
		return
	}
	mid := start + (end-start)>>1
	mergeSort(nums, start, mid, count)
	mergeSort(nums, mid+1, end, count)
	left, right := start, mid+1
	for left <= mid && right <= end {
		if nums[left] > 2*nums[right] {
			*count += mid - left + 1
			right++
		} else {
			left++
		}
	}
	left, right = start, mid+1
	tem := make([]int, end-start+1)
	index := 0
	for left <= mid && right <= end {
		if nums[left] < nums[right] {
			tem[index] = nums[left]
			index++
			left++
		} else {
			tem[index] = nums[right]
			index++
			right++
		}
	}
	for left <= mid {
		tem[index] = nums[left]
		index++
		left++
	}
	for right <= end {
		tem[index] = nums[right]
		index++
		right++
	}
	index = 0
	for i := start; i <= end; i++ {
		nums[i] = tem[index]
		index++
	}
}
